spacy/util.py

```diff
@@ -300,11 +302,9 @@ def find_matching_language(lang: str) -> Optional[str]:
     # is labeled that way is probably trying to be distinct from 'zh' and
     # shouldn't automatically match.
     match = langcodes.closest_supported_match(lang, possible_languages, max_distance=9)
-    if match == "mul":
-        # Convert 'mul' back to spaCy's 'xx'
-        return "xx"
-    else:
-        return match
+    if match is not None:
+        warnings.warn(Warnings.W124.format(lang=lang, renamed_lang=match))
+    return match
```

Comment:
To make this utility function easier to use (as in the tests), I think it would be better if it doesn't add any warnings internally and then the function that calls it can handle any unexpected values with warnings or errors, based on how they want to use the returned value.

Wherever the warning is, it shouldn't be the case that it warns on `("en", "en")` because the code wasn't modified.

Reply:
I've moved the warning to the `get_lang_class()` function which uses the `find_matching_language()`. It should only throw an error when `lang != match` is `True`, so that cases like `("en", "en")` don't throw an error.

Reply:
I'm still concerned that this way more warnings that we had before. We added `langcodes` adjustments without warnings for sensible cases and either no one has noticed or no one has minded the behavior, so re-introducing warnings seems odd in a lot of cases.

Is it possible to provide this error message for `xx` and `is` for all the CLI `-l` options (there are 4-5 cases total?) and then not for `get_lang_class` and `spacy.blank` so that they aren't any different than v3 for things like `en-US`?

Reply:
Yes, will implement is this way 😄

Reply:
I've removed the warnings from `get_lang_class()` and `spacy.blank`. The warning only gets thrown when `spacy convert` is used with either `-l xx` or `-l is`. When using `spacy init config -l xx` (or `is`) an error gets thrown that gives more information about that `xx` and `is` were renamed to `mul` and `isl` and that the user should change it accordingly.

I haven't found any other CLI commands that have the `-l` parameter except for `spacy init config` and `spacy convert`.

`spacy convert` only throws a warning instead of an error because most of the converters don't use the `lang` parameter anyway and would still work.

Reply:
Hmm, I may be confused about exactly what happens here, and it's something that's not really covered by the tests. Let me try it out more directly...

Reply:
The `--lang` setting is not actually used by any of the converters as far as I can tell.

Reply:
The error for `init config` shows up as a relatively ugly traceback, and I think it might be nicer if it were caught sooner and reported with `msg.fail` instead.

Reply:
Would it make sense to remove the warning from `spacy convert` then?

Reply:
Should the error for `init config` be specifically for `xx` and `is` cases?